src/util/taproot.rs

```diff
@@ -590,7 +590,7 @@ pub struct LeafInfo {
 
 impl LeafInfo {
     /// Creates an new [`LeafInfo`] from `script` and `ver` and no merkle branch.
-    fn new(script: Script, ver: LeafVersion) -> Self {
+    pub fn new(script: Script, ver: LeafVersion) -> Self {
```

Comment:
I am not super comfortable with this. I think we should not expose any constructors.

Reply:
`LeafInfo` can be constructed correctly using `NodeInfo` constructors(`new_leaf` or `new_leaf_with_ver`). Moreover, I don't see any use in exposing this. 
This is possibly dangerous because the user needs to make sure that Merkle proof path is correct.

Reply:
I just need a type having version, script and merkle path bound together downstream. I need it all the way around. If we can't have it here, I will not be able to reconstruct it, since everything here is private.

Reply:
You can call `control_block()` method on `TaprootSpendInfo` to get it. The iterator also outputs `LeafInfo` that you can use downstream. All it's members have accessor functions. 

My point is that we should only expose these after they are finalized. I don't want users creating misc leaves with Merkle proofs that don't verify. Let the high level take care of building it.

Reply:
Calling `control_block()` requires finalization, requiring internal key and secp object. Very annoying and destroying any normal API downstream.

Reply:
> The iterator also outputs LeafInfo that you can use downstream

The iterator works right and it has public accessor functions? You can fetch all the information from there and combine it in a custom type downstream.

If we have a new constructor that does not check if the leaf would be used correctly, we might as well expose all members as public fields.

Reply:
I had to develop my own tree types required for doing OP_RETURN taproot commitments, and those types also have iterators. I can't construct and return LeafNode from that iterators without it having constructors. Ok, I probably need to do another type for that and just provide convertor for it from this `LeafNode`. I agree that exposing it to others may be risky.